Allow override of bundler version or resolve from lock file

inverse opened the pull request:
Allow override of the default provided bundler from the image or resolve from `Gemfile.lock` if one exists.

inverse commented on entrypoint.sh:
```diff
@@ -8,6 +8,13 @@ if [ -n "$INPUT_PRE_BUILD_COMMANDS" ]; then
   eval "$INPUT_PRE_BUILD_COMMANDS"
 fi 
 
+echo "INPUT_BUNDLER_VERSON:${INPUT_BUNDLER_VERSON}"
```
For some reason this is coming back as empty when testing with:

```
      - uses: inverse/jekyll-action@override-bundler
        with:
          token: ${{ secrets.GITHUB_TOKEN }}
          bundler_version: '2.2.31'
          ```

helaili replied:
Replace `VERSON` with `VERSION` maybe?

inverse replied:
How did I miss that 🙈

inverse commented:
@helaili anything you would change on this?

helaili commented:
Sorry @inverse, time is a scarce resource. 
Would you mind adding a test case similar to [this one](https://github.com/helaili/jekyll-action/blob/master/.github/workflows/test.yml#L72-L117)?

inverse commented:
@helaili I added two cases cases where it would do what you outlined - I think it's correct?

I also added asserting the bundler version matches either explicitly or from the lockfile.

helaili commented on .github/workflows/test.yml:
```diff
@@ -69,6 +71,115 @@ jobs:
+    - uses: ./.github/actions/test
+      with: 
+        token: ${{ secrets.GITHUB_TOKEN }}
+        spec: cypress/integration/jekyll_src/**/*
+        suffix: "jekyll_src"
```
```suggestion
        spec: cypress/integration/jekyll_bundler_lock/**/*
        suffix: "jekyll_bundler_lock"
```

helaili reviewed (changes requested):
Thanks for working on this @inverse. I have only reviewed the test generation so far but I'd like it to be properly run before I review the rest of this PR.

inverse commented on .github/workflows/test.yml:
```diff
@@ -1,7 +1,9 @@
-  push
+  push:
+    branches: [master]
+  pull_request: ~
```
I think it's just for the default and is same as just having `pull_request:` We can revert it

inverse commented on .github/workflows/test.yml:
```diff
@@ -69,6 +71,115 @@ jobs:
+        target_branch: jekyll_src_test_pages_branch
+        jekyll_env: production
+      env:
+        JEKYLL_PAT: ${{ secrets.GITHUB_TOKEN }}
```
Do you mean like this

```

diff --git a/.github/workflows/test.yml b/.github/workflows/test.yml
index 7baa4fd..5d9f837 100644
--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -146,8 +146,8 @@ jobs:
         jekyll_src: sample_site
         target_branch: jekyll_bundler_lock_test_pages_branch
         jekyll_env: production
+        token: ${{ secrets.GITHUB_TOKEN }}
       env:
-        JEKYLL_PAT: ${{ secrets.GITHUB_TOKEN }}
         JEKYLL_DEBUG: true
     - id: extract-bundler-version
     ```

helaili commented on .github/workflows/test.yml:
```diff
@@ -69,6 +71,115 @@ jobs:
+        target_branch: jekyll_bundler_test_pages_branch
+        jekyll_env: production
+        bundler_version: 2.3.11
+        github-token: ${{secrets.JEKYLL_PAT}} 
```
My bad, sorry I got confused with the `github-script` action.
```suggestion
        token: ${{secrets.JEKYLL_PAT}} 
```

helaili commented on .github/workflows/test.yml:
```diff
@@ -69,6 +71,115 @@ jobs:
+        JEKYLL_DEBUG: true
+    - id: extract-bundler-version
+      run: echo ::set-output name=value::$(bundler -v | cut -c 16- | xargs)
+    - uses: nick-fields/assert-action@v1
+      with:
+        expected: 2.3.11
+        actual: ${{ steps.extract-bundler-version.outputs.value }}
+        comparison: exact
```
This step is failing. Please check https://github.com/helaili/jekyll-action/runs/6257218854?check_suite_focus=true.

inverse replied:
Just pushed something that should fix it, forgot to quote the set-output

helaili reviewed (changes requested):
Tests are not passing

helaili commented:
Tests are still failing. Trying to debug it, I've seen [this line](https://github.com/helaili/jekyll-action/runs/6272309590?check_suite_focus=true#step:5:13). 

Thing is `bundler` is being installed within the container, and the `echo "::set-output name=value::$(bundler -v | cut -c 16- | xargs)"` command is being called within the next step, so outside of the container where `bundler` is not installed or could be a totally different one. 

You need to run the `set-output` command from `entrypoint.sh` and referred to the value as coming from the `run` step. This step should actually be renamed for clarity.

inverse commented:
Ahh you're totall right :see_no_evil: 

> You need to run the set-output command from entrypoint.sh and referred to the value as coming from the run step. This step should actually be renamed for clarity.

What do you mean by that? I don't follow?

I  pushed a change which instead does the following:

- Adds the bundler version and exposes it on the page
- Modifies the cypress tests for the two additional flows to assert the bundler version

![image](https://user-images.githubusercontent.com/713525/166559503-8ed7baf0-d33e-4397-9490-a969b53432a6.png)

helaili commented on .github/workflows/test.yml:
```diff
@@ -69,6 +71,101 @@ jobs:
+        
+  jekyll_bundler_from_lock_publish: 
+    runs-on: ubuntu-latest
+    needs: [basic_test, jekyll_bundler_from_lock_run]
```
```suggestion
    needs: [jekyll_bundler_test, jekyll_bundler_from_lock_run]
```

helaili reviewed (changes requested):
I think it's getting closed. We can unfortunately publish and test in sequence as there is only one published site at a time, so the publish needs to happen after the last test and then we run the test.

inverse commented:
@helaili good catch! how does the graph look now? https://github.com/helaili/jekyll-action/actions/runs/2296309955

inverse commented:
@helaili I think I know why...

So this test case is when we force install a newer version of bundler than what the application is bundled with. Yet the system has `2.1.4` so it bundles with that version as it's in the lockfile :/

Not sure what a good test would be for this case? 


I could remove that part and we just install the bundler version from the lockfile which probably covers the majority of cases...

helaili commented:
I'm removing the Gemfile.lock file during the `jekyll_bundler_run` job so that it is generated with the right `Bundler` version. I'm also now relying on the `bundle -v` command as opposed to `bundler -v` to figure out the version number as `bundler is not in the path in this specific case. 

Thanks for this contribution @inverse

inverse commented:
@helaili That's for taking the time to explain :+1:
